File: em_mixed_gaussian_project.r
library(ggplot2)

generate_samples <- function(
    mixing_coefficients,
    means,
    standard_deviations,
    number_of_samples = 1000) {


  # should check that lengths are equal

  data <- c()
  choices_count <- rep(0, length(mixing_coefficients))

  # select one of the gaussian distributions according
  # to the mixing coefficients
  choices <- sample(
                    x = seq_along(mixing_coefficients),
                    size = number_of_samples,
                    prob = mixing_coefficients,
                    replace = TRUE)

  # let us see that the number of selections make sense
  for (i in 1:number_of_samples){
    choices_count[choices[i]] <- choices_count[choices[i]] + 1
  }
  print(choices_count)

  # for each selection, sample from the gaussian
  data <- c(
            data,
            rnorm(
                  n = number_of_samples,
                  mean = means[choices],
                  sd = standard_deviations[choices]))

  # plot histogram and gaussian curves
  df <- data.frame(data)

  p <- ggplot(df, aes(x = data)) +
    geom_histogram(
                   aes(y = ..density..),
                   bins = 30,
                   fill = "blue",
                   alpha = 0.6) +
    stat_function(fun = function(x) {
      Reduce(`+`, lapply(1:length(means), function(i) {
        dnorm(
              x = x,
              mean = means[i],
              sd = standard_deviations[i]) * mixing_coefficients[i]
      }))
    }, color = "red") +
    labs(title = "Histogram of Mixture of Gaussians",
         x = "Value",
         y = "Density") +
    theme_bw()

  return(data)
}


initial_values_kmeans <- function(data, k = 3) {
  set.seed(42)  # For reproducibility

  # Randomly select k initial centroids
  centroids <- sample(x = data, size = k, replace = FALSE)

  # Initialize empty clusters
  clusters <- vector(mode = "list", length = k)

  for (iteration in 1:100) {
    # Reset clusters
    clusters <- vector(mode = "list", length = k)

    # Assign each data point to the closest centroid
    for (item in data) {
      distances <- abs(item - centroids)
      closest_centroid <- which.min(distances)
      clusters[[closest_centroid]] <- c(clusters[[closest_centroid]], item)
    }

    # Compute new centroids
    new_centroids <- sapply(1:k, function(i) {
      if (length(clusters[[i]]) > 0) {
        mean(clusters[[i]])
      } else {
        centroids[i]  # Keep old centroid if no points are assigned
      }
    })

    # Check for convergence
    if (max(abs(new_centroids - centroids)) < 0.0001) {
      break
    }

    centroids <- new_centroids
  }

  # Compute standard deviations
  clusters_standard_devs <- sapply(1:k, function(cluster_idx) {
    if (length(clusters[[cluster_idx]]) > 0) {
      sqrt(sum((clusters[[cluster_idx]] - centroids[cluster_idx])^2) / length(clusters[[cluster_idx]]))
    } else {
      0
    }
  })

  # Compute mixing coefficients for each cluster
  mixing_coefficients <- sapply(1:k, function(cluster_idx) {
    length(clusters[[cluster_idx]]) / length(data)
  })

  return(list(
              centroids = centroids,
              standard_devs = clusters_standard_devs,
              mixing_coefficients = mixing_coefficients))
}


log_likelihood <- function(data, means, standard_deviations, mixing_coefficients) {

  number_of_gaussians <- length(mixing_coefficients)

  # Initialize likelihood matrix
  likelihood <- matrix(0, nrow = length(data), ncol = number_of_gaussians)

  # Compute likelihood for each Gaussian component
  for (k in 1:number_of_gaussians) {
    likelihood[, k] <- mixing_coefficients[k] * dnorm(data, mean = means[k], sd = standard_deviations[k])
  }

  # Compute the log-likelihood
  log_likelihood_value <- sum(log(rowSums(likelihood)))

  return(log_likelihood_value)
}


expectation_step <- function(data, means, standard_deviations, mixing_coefficients) {

  number_of_gaussians <- length(mixing_coefficients)

  # Initialize gamma matrix
  gamma <- matrix(0, nrow = length(data), ncol = number_of_gaussians)

  # Calculate denominator (total probability for each data point)
  den_total <- 0
  for (k in 1:number_of_gaussians) {
    den_total <- den_total + mixing_coefficients[k] * dnorm(data, mean = means[k], sd = standard_deviations[k])
  }

  # Calculate numerator and compute gamma
  for (k in 1:number_of_gaussians) {
    gamma[, k] <- (mixing_coefficients[k] * dnorm(data, mean = means[k], sd = standard_deviations[k])) / den_total
  }

  return(gamma)
}


maximization_step <- function(data, gamma, means, standard_deviations, mixing_coefficients) {

  no_gaussians <- length(mixing_coefficients)
  m <- length(data)

  # Compute cluster responsibilities
  m_c <- colSums(gamma)  # Sum of responsibilities for each Gaussian

  # Compute new mixing coefficients
  new_mixing_coefficients <- m_c / m

  # Initialize new means and standard deviations
  new_means <- numeric(no_gaussians)
  new_standard_deviations <- numeric(no_gaussians)

  # Compute new means and standard deviations for each Gaussian
  for (k in 1:no_gaussians) {
    new_means[k] <- sum(gamma[, k] * data) / m_c[k]
    new_standard_deviations[k] <- sqrt(sum(gamma[, k] * (data - means[k])^2) / m_c[k])
  }

  return(list(
    means = new_means,
    standard_devs = new_standard_deviations,
    mixing_coefficients = new_mixing_coefficients
  ))
}



mixing_coefficients <- c(0.2, 0.5, 0.3)
means <- c(6, 0, -7)
standard_deviations <- c(2, 1, 1.5)

# generate the data
data <- generate_samples(mixing_coefficients, means, standard_deviations)

# initial values
kmeans_ret <- initial_values_kmeans(data)


centroids <- kmeans_ret$centroids
standard_devs <- kmeans_ret$standard_devs
mixing_coefficients <- kmeans_ret$mixing_coefficients

print("initial values")
print(centroids)
print(standard_devs)
print(mixing_coefficients)


log_likelihoods <- c()
for (i in 1:100) {
  gamma <- expectation_step(data, centroids, standard_devs, mixing_coefficients)

  max_ret <- maximization_step(data, gamma, centroids, standard_devs, mixing_coefficients)

  centroids <- max_ret$means
  standard_devs <- max_ret$standard_devs
  mixing_coefficients <- max_ret$mixing_coefficients 

  ll <- log_likelihood(data, means, standard_devs, mixing_coefficients)

  log_likelihoods <- c(log_likelihoods, ll)

  if (i>2 &&  abs(ll - log_likelihoods[i-1]) < 1e-5) {
    print("converged")
    break
  }

  # form string
  iteration_result_str <- paste( "iteration: ", i)
  for (j in 1:length(centroids)) {
    iteration_result_str <- paste(iteration_result_str, " centroid: ", centroids[j])
  }
  for (j in 1:length(standard_devs)) {
    iteration_result_str <- paste(iteration_result_str, " standard_dev: ", standard_devs[j])
  }
  for (j in 1:length(mixing_coefficients)) {
    iteration_result_str <- paste(iteration_result_str, " mixing_coefficient: ", mixing_coefficients[j])
  }
  print(iteration_result_str)

}

File: em_mixed_gaussian_project.py
import numpy as np
import random
import scipy.stats
from scipy.stats import norm
import matplotlib.pyplot as plt

def log_likelihood(data, means, standard_deviations, mixing_coefficients):

    number_of_gaussians = len(mixing_coefficients)

    likelihood = np.zeros((len(data), number_of_gaussians), dtype=float)

    for k in range(number_of_gaussians):
        likelihood[:, k] =  mixing_coefficients[k] * norm(means[k], standard_deviations[k]).pdf(data)

    log_likelihood = np.sum(np.log(np.sum(likelihood, axis=1)))
    
    return log_likelihood


def expectation_step(data, means, standard_deviations, mixing_coefficients):
    
    number_of_gaussians = len(mixing_coefficients)
    
    gamma = np.zeros((len(data), number_of_gaussians), dtype=float)
    
    
    # calculate denominator
    den_total = 0
    for k in range(number_of_gaussians):
        den_total += mixing_coefficients[k] * norm(means[k], standard_deviations[k]).pdf(data)
    
    # calculate the numerator
    for k in range(number_of_gaussians):
        gamma[:, k] = mixing_coefficients[k] * norm(means[k], standard_deviations[k]).pdf(data) / den_total
        
    return gamma



def maximization_step(data, gamma, means, standard_deviations, mixing_coefficients):

    no_gaussians = len(mixing_coefficients)
    m = len(data)
    new_means = np.zeros(shape=(means.shape))
    new_standard_deviations = np.zeros(shape=(standard_deviations.shape))

    m_c = np.sum(gamma, axis=0)

    new_mixing_coefficients = m_c / m

    for k in range(no_gaussians):
        new_means[k] = np.sum(gamma[:, k] * data) / m_c[k]
        new_standard_deviations[k] = np.sqrt(np.sum(gamma[:, k] * (data - means[k]) ** 2) / m_c[k])

    return new_means, new_standard_deviations, new_mixing_coefficients


def simulate_points(means, standard_deviations, mixing_coefficients, number_of_points=1000):
    '''
    Simulate 1000 data points from a mixture of Gaussians
    '''

    data = []

    counts = np.zeros(len(mixing_coefficients))
    choices = np.arange(len(mixing_coefficients))

    for _ in range(number_of_points):
        choice = random.choices(choices, mixing_coefficients)[0]
        counts[choice] += 1
        data.append(float(random.gauss(means[choice], standard_deviations[choice])))

    print("Counts: ", counts)

    plt.hist(data, bins=30, density=True, alpha=0.6, color='g')
    x_values = np.linspace(min(data), max(data), number_of_points)
    for idx, mix_coeff in enumerate(mixing_coefficients):
        plt.plot(
            x_values,
            mix_coeff * norm(means[idx], standard_deviations[idx]).pdf(x_values),
            label=f'Gaussian {idx+1}')
    plt.xlabel("Value")
    plt.ylabel("Probability Density")
    plt.title(f"Simulated Data (coeff = {mixing_coefficients}, means = {means}, std = {standard_deviations})")
    plt.legend()
    plt.show()

    print(data)

    return data

def main():
    data = [7.33, 9.66, 9.00, 5.00, 8.82, 10.36, 1.78, 8.42, 10.20, 7.69, 1.57, 10.85,8.66, 8.18,
            -1.28, 3.06, 10.18, 2.86, 10.90, 10.80, 10.00, 1.46, 10.16, 1.91, 1.92, 8.47, 4.17,
            3.75, -0.02, 10.53 ]

    number_of_gaussians = 2
    max_iterations = 50
    convergence_value = 0.0001


    random.seed(80)

    means = np.array([random.random(), random.random()])
    standard_deviations = np.array([random.random(), random.random()])

    # can work with single variable and do 1- for the other, but i want to reuse
    mixing_coefficients = np.zeros(number_of_gaussians)
    mixing_coefficients[0] = 1 / number_of_gaussians
    mixing_coefficients[1] = 1 / number_of_gaussians

    print("Initial Values: ")
    print("means = ", means)
    print("standard_deviations = ", standard_deviations)
    print("mixing_coefficients: ", mixing_coefficients)
    
    
    log_likelihoods = []
    
    for i in range(max_iterations):
        gamma = expectation_step(data, means, standard_deviations, mixing_coefficients)
        means, standard_deviations, mixing_coefficients = maximization_step(data, gamma, means, standard_deviations, mixing_coefficients)
        log_likelihoods.append(log_likelihood(data, means, standard_deviations, mixing_coefficients))
        
        if i>1 and log_likelihoods[-1] - log_likelihoods[-2] < convergence_value:
            break
        

    print("Updated Values: ")
    print("means = ", means)
    print("standard_deviations = ", standard_deviations)
    print("mixing_coefficients: ", mixing_coefficients)
    print("Log likelihood: ", log_likelihoods[-1])
    
    
    # plt.plot(log_likelihoods[1:], label="Log likelihood")
    # plt.xlabel('Iteration')
    # plt.ylabel('Log likelihood')
    # plt.show()
    
    
    # plot histogram scale gaussians with mixing coefficients
    plt.hist(data, bins=6, density=True, alpha=0.6, color='g')
    x_values = np.linspace(min(data), max(data), 1000)
    for k in range(number_of_gaussians):
        plt.plot(x_values, norm(means[k], standard_deviations[k]).pdf(x_values), label=f'Gaussian {k+1}')
        
    plt.xlabel("Value")
    plt.ylabel("Probability Density")
    plt.title(f"EM Algorithm - Gaussian Mixture Model (GMM) - {number_of_gaussians} Gaussians")
    plt.legend()
    plt.show()
    
    
def initial_values_knn(data, k=3):

    centroids = np.random.choice(data, k, replace=False)

    clusters = [[] for _ in range(k)]
    print("Clusters: ", clusters)

    for _ in range(100):
        for item in data:
            closest_centroid = np.argmin([abs(item - c) for c in centroids])
            clusters[closest_centroid].append(item)
            
        new_centroids = np.array([np.mean(cluster) if cluster else centroids[i] for i, cluster in enumerate(clusters)])
        
        if np.max(np.abs(new_centroids - centroids)) < 0.0001:
            break
            
        centroids = new_centroids
    
    standard_devs = []
    for i in range(k):
        mean = centroids[i]
        cluster_data = clusters[i]
        standard_devs.append(float(np.sqrt(np.sum((cluster_data - mean) ** 2) / len(cluster_data))))
    
    scaling_factors = [float(1/(np.sqrt(2*np.pi)*sd)) for sd in standard_devs]
    
    return centroids, np.array(standard_devs), scaling_factors


if __name__ == "__main__":

    means = np.array([2,11,-5])
    standard_deviations = np.array([1,2,0.5])
    mixing_coefficients = np.array([0.3,0.3,0.4])

    data = simulate_points(means, standard_deviations, mixing_coefficients)
    
    
    means, standard_deviations, mixing_coefficients = initial_values_knn(data, k=3)
    
    print(means)
    print(standard_deviations)
    print(mixing_coefficients)
    
    print("Initial Values: ")
    print("means = ", means)
    print("standard_deviations = ", standard_deviations)
    print("mixing_coefficients: ", mixing_coefficients)
    
    
    log_likelihoods = []
    
    for i in range(1000):
        gamma = expectation_step(data, means, standard_deviations, mixing_coefficients)
        means, standard_deviations, mixing_coefficients = maximization_step(data, gamma, means, standard_deviations, mixing_coefficients)
        log_likelihoods.append(log_likelihood(data, means, standard_deviations, mixing_coefficients))
        
        if i>1 and log_likelihoods[-1] - log_likelihoods[-2] < 0.00001:
            break
        

    print("Updated Values: ")
    print("means = ", means)
    print("standard_deviations = ", standard_deviations)
    print("mixing_coefficients: ", mixing_coefficients)
    print("Log likelihood: ", log_likelihoods[-1])